src/base/RefCount.h

```diff
@@ -87,6 +87,14 @@ class RefCount
         return p.p_ != p_;
     }
 
+    bool operator == (C const * p) const {
+        return p == p_;
+    }
+
+    bool operator != (C const * p) const {
+        return p!= p_;
+    }
```

Comment:
There are many ways to fix ESI build. To find the overall best solution, let's ignore ESI build problems for a second, especially since the proposed fix applies to the _entire_ Squid code.

> There's some comparisons that implicitly construct a RefCounted object from a pointer

This PR manually adds "simpler" code that an optimizing compiler, in theory, may be able to generate itself. The question is: Are modern compilers smart enough to optimize away creation of a temporary RefCount object?

* If yes, then we are writing (trickier than it may seem!) code that the compiler can generate for us. We should not do that.
* If no, then we should try to cover more raw pointer equality cases, similar to what we did for RefCount<C> creation from a raw pointer -- `C` in these new equality test operators should probably be a templated Other parameter.

Do you know (or can you find out experimentally) the answer to the above question?

Reply:
> The question is: Are modern compilers smart enough to optimize away creation of a temporary RefCount object?

I don't think the issue here is whether an optimizing compiler could optimize away the creation of a temporary RefCount object, it's that there's ambiguity about which of two candidates to use.  I believe this is due to how comparison operators [can be reversed/rewritten in C++20](https://stackoverflow.com/a/65947030).  We could also modify ESI.cc, as you mentioned.

Reply:
> > Are modern compilers smart enough to optimize away creation of a temporary RefCount object?
 
> the issue here is ... that there's ambiguity about which of two candidates to use.

AFAICT, you are thinking about a specific ESI-related build failure. I am not disagreeing with you, but let's ignore that specific build failure for a moment. Assume, for the sake of the argument, that ESI code does not exist.

> I believe this is due to how comparison operators [can be reversed/rewritten in C++20](https://stackoverflow.com/a/65947030).

Yes, we share that understanding. I am just trying to put that [ESI-introduced/specific] ambiguity aside (for a moment) when deciding whether/how we should improve RefCount. After all, all the correct prose aside, this PR is actually changing RefCount, not any ESI-specific code.

Reply:
I see your point.  It seems the less invasive change is to just fix the couple places in Esi.cc where this issue occurs by using `.getRaw()` in the comparison, if you think that is a reasonable approach.

Reply:
> Alex: Are modern compilers smart enough to optimize away creation of a temporary RefCount object?

I built a [test](https://godbolt.org/z/rovMhfW1x) to answer my own question: The short answer is "no". Inlining RefCount shows that optimizing GCC v10.3:

* Avoids temporaries when comparing Pointer with nullptr (function test1() at the bottom of the test code), but 
* Cannot avoid temporaries when comparing Pointer with an arbitrary raw pointer (function test2() at the bottom of the test code). Search for "XXX: Added for testing" to remove your RefCount methods and see test2() assembly on the right exploding from a trivial 2-line pointer comparison to countless lines creating complex temporary objects.

Same result for GCC v12 and clang.

Thus, the proposed RefCount changes are an optimization that should be welcomed (regardless of ESI and C++20 issues). I am resolving this thread, adjusting PR title/description, and will post specific improvement suggestions.

Reply:
Thanks Alex, I appreciate it!